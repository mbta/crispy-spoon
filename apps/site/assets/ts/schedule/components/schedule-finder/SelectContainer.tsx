import React, { ReactElement, ReactNode } from "react";
import { handleReactEnterKeyPress } from "../../../helpers/keyboard-events-react";
import renderSvg from "../../../helpers/render-svg";
import arrowIcon from "../../../../static/images/icon-down-arrow.svg";

interface SelectContainerProps {
  children: ReactNode;
  error?: boolean;
  handleClick?: Function;
}

const SelectContainer = ({
  children,
  error = false,
  handleClick
}: SelectContainerProps): ReactElement<HTMLElement> => (
  <div
    tabIndex={0}
    className={`c-select-custom__container ${error ? "error" : ""}`}
    role="button"
    onClick={e => {
      if (handleClick) {
        handleClick(e);
      }
    }}
    onKeyUp={e =>
      handleReactEnterKeyPress(e, () => {
        if (handleClick) {
          handleClick(e);
        }
      })
    }
  >
    {children}
    {renderSvg("c-svg__icon c-select-custom__arrow", arrowIcon, true)}
  </div>
);

export default SelectContainer;
