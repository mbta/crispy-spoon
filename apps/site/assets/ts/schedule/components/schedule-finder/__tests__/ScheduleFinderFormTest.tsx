import React from "react";
import { shallow, mount } from "enzyme";
import { Route } from "../../../../__v3api";
import { SimpleStopMap } from "../../__schedule";
import ScheduleFinderForm from "../ScheduleFinderForm";
import SelectContainer from "../SelectContainer";

jest.mock("../../../../helpers/use-fetch", () => ({
  __esModule: true,
  hasData: () => false,
  isLoading: () => true,
  isNotStarted: () => false,
  default: jest.fn().mockImplementation(() => [{ status: 2 }, jest.fn()])
}));

const route: Route = {
  description: "",
  direction_destinations: { 0: "Oak Grove", 1: "Forest Hills" },
  direction_names: { 0: "Inbound", 1: "Outbound" },
  id: "Orange",
  long_name: "Orange Line",
  name: "Orange",
  type: 1,
  line_id: null
};

const oneDirectionRoute: Route = {
  description: "",
  direction_destinations: { 0: "Destination", 1: null },
  direction_names: { 0: "Outbound", 1: null },
  id: "route",
  long_name: "the route",
  name: "Route",
  type: 1,
  line_id: null
};

const stops: SimpleStopMap = {
  "1": [
    {
      name: "SL",
      id: "741",
      is_closed: false,
      zone: "1"
    },
    {
      name: "Abc",
      id: "123",
      is_closed: false,
      zone: null
    },
    {
      name: "Def",
      id: "456",
      is_closed: false,
      zone: null
    },
    {
      name: "Wellington",
      id: "place-welln",
      is_closed: true,
      zone: null
    }
  ],
  "0": [
    {
      name: "Wellington",
      id: "place-welln",
      is_closed: true,
      zone: null
    },
    {
      name: "Abc",
      id: "123",
      is_closed: false,
      zone: null
    },
    {
      name: "SL",
      id: "741",
      is_closed: false,
      zone: "1"
    }
  ]
};

const noCall = () => {
  throw new Error("should not have been called");
};

describe("ScheduleFinderForm", () => {
  it("matches snapshot", () => {
    const wrapper = shallow(
      <ScheduleFinderForm
        onDirectionChange={noCall}
        onOriginChange={noCall}
        onOriginSelectClick={noCall}
        onSubmit={noCall}
        route={route}
        selectedDirection={0}
        selectedOrigin={null}
        stopsByDirection={stops}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it("includes only valid directions in the direction picker", () => {
    const wrapper = shallow(
      <ScheduleFinderForm
        onDirectionChange={noCall}
        onOriginChange={noCall}
        onOriginSelectClick={noCall}
        onSubmit={noCall}
        route={oneDirectionRoute}
        selectedDirection={0}
        selectedOrigin={null}
        stopsByDirection={stops}
      />
    );

    const directions = wrapper
      .find("select")
      .first()
      .find("option");
    expect(directions).toHaveLength(1);
    expect(directions.text()).toContain("OUTBOUND");
  });

  it("shows an error if the form is submitted without an origin", () => {
    const submitted = jest.fn();
    const wrapper = shallow(
      <ScheduleFinderForm
        onDirectionChange={noCall}
        onOriginChange={noCall}
        onOriginSelectClick={noCall}
        onSubmit={submitted}
        route={route}
        selectedDirection={0}
        selectedOrigin={null}
        stopsByDirection={stops}
      />
    );

    wrapper.find("form").simulate("submit", { preventDefault: () => {} });

    expect(wrapper.text()).toContain("Please provide an origin");
    expect(submitted).not.toHaveBeenCalled();
    expect(
      wrapper
        .find(SelectContainer)
        .last()
        .props().error
    ).toEqual(true);
  });

  it("calls the submit handler and clears the error", () => {
    const submitted = jest.fn();
    const wrapper = shallow(
      <ScheduleFinderForm
        onDirectionChange={noCall}
        onOriginChange={noCall}
        onOriginSelectClick={noCall}
        onSubmit={submitted}
        route={route}
        selectedDirection={0}
        selectedOrigin={null}
        stopsByDirection={stops}
      />
    );

    // Show the error first
    wrapper.find("form").simulate("submit", { preventDefault: () => {} });
    wrapper.setProps({ selectedOrigin: "123" });
    wrapper.find("form").simulate("submit", { preventDefault: () => {} });

    expect(wrapper.text()).not.toContain("Please provide an origin");
    expect(submitted).toHaveBeenCalledTimes(1);
  });

  it("calls the origin select handler and clears the error", () => {
    const originClicked = jest.fn();
    const wrapper = shallow(
      <ScheduleFinderForm
        onDirectionChange={noCall}
        onOriginChange={noCall}
        onOriginSelectClick={originClicked}
        route={route}
        selectedDirection={0}
        selectedOrigin={null}
        stopsByDirection={stops}
      />
    );

    // Show the error first
    wrapper.find("form").simulate("submit", { preventDefault: () => {} });
    // Click on the SelectContainer for the origin select
    wrapper
      .find("SelectContainer")
      .last()
      // @ts-ignore -- types for `invoke` are too restrictive?
      .invoke("handleClick")();

    expect(wrapper.text()).not.toContain("Please provide an origin");
    expect(originClicked).toHaveBeenCalledTimes(1);
  });

  it("calls the direction and origin change handlers", () => {
    const directionChanged = jest.fn();
    const originChanged = jest.fn();
    const wrapper = shallow(
      <ScheduleFinderForm
        onDirectionChange={directionChanged}
        onOriginChange={originChanged}
        onOriginSelectClick={noCall}
        onSubmit={noCall}
        route={route}
        selectedDirection={0}
        selectedOrigin={null}
        stopsByDirection={stops}
      />
    );

    wrapper
      .find("select")
      .first()
      .simulate("change", { target: { value: 1 } });
    wrapper
      .find("select")
      .last()
      .simulate("change", { target: { value: "123" } });

    expect(directionChanged).toHaveBeenCalledWith(1);
    expect(originChanged).toHaveBeenCalledWith("123");
  });

  it("detects click and keyUp events in SelectContainer elements", () => {
    const originSpy = jest.fn();

    const wrapper = mount(
      <ScheduleFinderForm
        onDirectionChange={() => {}}
        onOriginChange={() => {}}
        onOriginSelectClick={originSpy}
        onSubmit={noCall}
        route={route}
        selectedDirection={0}
        selectedOrigin={null}
        stopsByDirection={stops}
      />
    );

    // detect click event:
    wrapper
      .find(SelectContainer)
      .at(1)
      .simulate("click");

    expect(originSpy).toHaveBeenCalledTimes(1);

    // detect keyUp event:
    originSpy.mockRestore();

    wrapper
      .find(".c-select-custom__container")
      .at(1)
      .simulate("keyUp", { key: "Enter" });

    expect(originSpy).toHaveBeenCalledTimes(1);

    originSpy.mockRestore();
    wrapper.unmount();
  });
});
