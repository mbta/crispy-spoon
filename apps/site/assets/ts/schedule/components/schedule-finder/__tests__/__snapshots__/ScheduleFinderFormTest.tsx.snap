// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScheduleFinderForm matches snapshot 1`] = `
<Fragment>
  <form
    onSubmit={[Function]}
  >
    <h2
      className="schedule-finder__heading"
    >
      <span
        aria-hidden="true"
        className="notranslate c-svg__icon"
        dangerouslySetInnerHTML={
          Object {
            "__html": "SVG",
          }
        }
      />
       Schedule Finder
    </h2>
    <div
      className="schedule-finder__prompt"
    >
      Choose a stop to get schedule information and real-time departure predictions.
    </div>
    <div
      className="schedule-finder__inputs"
    >
      <label
        className="schedule-finder__label"
      >
        Choose a direction
        <SelectContainer>
          <select
            className="c-select-custom notranslate"
            data-testid="schedule-finder-direction-select"
            onChange={[Function]}
            value={0}
          >
            <option
              aria-label="INBOUND 
  Oak Grove"
              key="0"
              value={0}
            >
              INBOUND 
  Oak Grove
            </option>
            <option
              aria-label="OUTBOUND 
  Forest Hills"
              key="1"
              value={1}
            >
              OUTBOUND 
  Forest Hills
            </option>
          </select>
        </SelectContainer>
      </label>
      <label
        className="schedule-finder__label"
      >
        Choose an origin stop
        <SelectContainer
          error={false}
          handleClick={[Function]}
        >
          <select
            className="c-select-custom c-select-custom--noclick notranslate"
            onChange={[Function]}
            value=""
          >
            <option
              value=""
            >
              Select
            </option>
            <option
              aria-label="Wellington"
              key="place-welln"
              value="place-welln"
            >
              Wellington
            </option>
            <option
              aria-label="Abc"
              key="123"
              value="123"
            >
              Abc
            </option>
            <option
              aria-label="SL"
              key="741"
              value="741"
            >
              SL
            </option>
          </select>
        </SelectContainer>
      </label>
    </div>
    <div
      className="schedule-finder__submit text-right"
    >
      <input
        className="btn btn-primary"
        type="submit"
        value="Get schedules"
      />
    </div>
  </form>
</Fragment>
`;
